lib/matplotlib/collections.py

```diff
@@ -1106,78 +1106,134 @@ def legend_elements(self, prop="colors", num="auto",
         """
         handles = []
         labels = []
-        hasarray = self.get_array() is not None
-        if fmt is None:
-            fmt = mpl.ticker.ScalarFormatter(useOffset=False, useMathText=True)
-        elif isinstance(fmt, str):
-            fmt = mpl.ticker.StrMethodFormatter(fmt)
-        fmt.create_dummy_axis()
 
         if prop == "colors":
-            if not hasarray:
+            arr = self.get_array()
+            if arr is None:
                 warnings.warn("Collection without array used. Make sure to "
                               "specify the values to be colormapped via the "
                               "`c` argument.")
                 return handles, labels
-            u = np.unique(self.get_array())
-            size = kwargs.pop("size", mpl.rcParams["lines.markersize"])
+            _size = kwargs.pop("size", mpl.rcParams["lines.markersize"])
+
+            def _get_color_and_size(value):
+                return self.cmap(self.norm(value)), _size
+
         elif prop == "sizes":
-            u = np.unique(self.get_sizes())
-            color = kwargs.pop("color", "k")
+            arr = self.get_sizes()
+            _color = kwargs.pop("color", "k")
+
+            def _get_color_and_size(value):
```

Comment:
I'm not sure I like the idea of different defs for the same function, and that function depending on the scope above.  I suppose it works, but its easy to see someone later getting confused and breaking it.

Reply:
Functions depending on the outside scope is done everywhere, see the use of `warnings.warn` just a few lines up, and I don't see why `arr` is ok but `_get_color_and_size` is not?

This is the clearest version I've come up with. We remove a bit of duplicated code by concentrating all the prop checks to 1 place which makes it much easier in my opinion to see what happens when changing that parameter.

Reply:
If `_color` is changed between here and where `_get_color_and_size` is called, it is not clear that it is being used by this method, and bad bugs result. 

I'm not trying to claim that Matplotlib code is devoid of indirection like this, but IMHO it is generally a bad idea.  OTOH maybe I'm wrong at its good practice; I don't claim to be an authority on good python programming practice.

Reply:
Seems like standard risks using mutable objects to me, similar issues can occur with `arr` as well by doing `arr[:]=42` somewhere before it's used.

Do you have a suggestion in mind? Because the alternatives I see aren't as clear or duplicates code.

Reply:
My suggestion would be to "duplicate" the code to keep it explicit.

Reply:
@jklymak, how about this version? I think this addresses your concern.